File: Meeting#2/python/quick_find.py
class QuickFind:

    def __init__(self, size):
        self._ids = list(range(0, size))
        self._count = size
        pass

    def connected(self, p: int, q: int) -> bool:
        return self.find(p) == self.find(q)

    def union(self, p: int, q: int) -> None:
        if p == q:
            return
        p_id = self.find(p)
        q_id = self.find(q)
        if p_id == q_id:
            return
        for i in range(0, len(self._ids)):
            if self._ids[i] == p_id:
                self._ids[i] = q_id
        self._count -= 1

    def count(self) -> int:
        return self._count

    def find(self, p: int) -> int:
        if p < 0 or p >= len(self._ids):
            raise ValueError(f"P value {p} isn't between 0 and {len(self._ids)}")
        return self._ids[p]

File: Meeting#2/java/src/QuickFind.java
public class QuickFind implements UnionFind {

    private final int[] id;
    private int count;

    public QuickFind(final int size) {
        id = new int[size];
        for (int i = 0; i < size; i++) {
            id[i] = i;
        }
        count = size;
    }

    @Override
    public boolean connected(int p, int q) {
        return find(p) == find(q);
    }

    @Override
    public void union(int p, int q) {
        if (p == q) {
            return;
        }
        final int pId = find(p);
        final int qId = find(q);
        if (pId == qId) {
            return;
        }
        for (int i = 0; i < id.length; i++) {
            if (id[i] == pId) {
                id[i] = qId;
            }
        }
        count--;
    }

    @Override
    public int count() {
        return count;
    }

    @Override
    public int find(int p) {
        final int length = id.length;
        if (p < 0 || p >= length) {
            throw new IllegalArgumentException(String.format("P value %1$d isn't between 0 and %2$d", p, length));
        }
        return id[p];
    }

}